File: backend/app/api/portfolio.py
from typing import Any, Dict, List

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.database_models import Portfolio, Position, Trade
from app.services.bitvavo_api_secure import BitvavoAPI

router = APIRouter()


# Pydantic models
class PortfolioCreate(BaseModel):
    name: str
    description: str = ""


class PortfolioResponse(BaseModel):
    id: int
    name: str
    description: str
    total_value: float
    is_active: bool


class PositionResponse(BaseModel):
    id: int
    symbol: str
    quantity: float
    average_price: float
    current_price: float
    unrealized_pnl: float


class TradeResponse(BaseModel):
    id: int
    symbol: str
    side: str
    order_type: str
    quantity: float
    price: float
    filled_quantity: float
    filled_price: float
    status: str
    created_at: str


@router.post("/", response_model=PortfolioResponse)
async def create_portfolio(
    portfolio: PortfolioCreate, db: Session = Depends(get_db)
) -> PortfolioResponse:
    """Create a new portfolio"""
    try:
        # TODO: Add user authentication and get actual user_id
        db_portfolio = Portfolio(
            user_id=1,  # Placeholder
            name=portfolio.name,
            description=portfolio.description,
        )
        db.add(db_portfolio)
        db.commit()
        db.refresh(db_portfolio)

        return PortfolioResponse(
            id=db_portfolio.id,  # type: ignore
            name=db_portfolio.name,  # type: ignore
            description=db_portfolio.description or "",  # type: ignore
            total_value=db_portfolio.total_value,  # type: ignore
            is_active=db_portfolio.is_active,  # type: ignore
        )
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/", response_model=List[PortfolioResponse])
async def get_portfolios(db: Session = Depends(get_db)) -> List[PortfolioResponse]:
    """Get all portfolios for the user"""
    try:
        # TODO: Add user authentication and filter by actual user_id
        portfolios = db.query(Portfolio).filter(Portfolio.user_id == 1).all()

        result = []
        for portfolio in portfolios:
            result.append(
                PortfolioResponse(
                    id=portfolio.id,  # type: ignore
                    name=portfolio.name,  # type: ignore
                    description=portfolio.description or "",  # type: ignore
                    total_value=portfolio.total_value,  # type: ignore
                    is_active=portfolio.is_active,  # type: ignore
                )
            )

        return result
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{portfolio_id}", response_model=PortfolioResponse)
async def get_portfolio(
    portfolio_id: int, db: Session = Depends(get_db)
) -> PortfolioResponse:
    """Get a specific portfolio"""
    try:
        portfolio = (
            db.query(Portfolio)
            .filter(
                Portfolio.id == portfolio_id,
                Portfolio.user_id == 1,  # TODO: Use actual user_id
            )
            .first()
        )

        if not portfolio:
            raise HTTPException(status_code=404, detail="Portfolio not found")

        return PortfolioResponse(
            id=portfolio.id,  # type: ignore
            name=portfolio.name,  # type: ignore
            description=portfolio.description or "",  # type: ignore
            total_value=portfolio.total_value,  # type: ignore
            is_active=portfolio.is_active,  # type: ignore
        )
    except HTTPException:
        # Re-raise HTTP exceptions (like 404) without wrapping them
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{portfolio_id}/positions", response_model=List[PositionResponse])
async def get_portfolio_positions(
    portfolio_id: int, db: Session = Depends(get_db)
) -> List[PositionResponse]:
    """Get all positions in a portfolio"""
    try:
        positions = (
            db.query(Position).filter(Position.portfolio_id == portfolio_id).all()
        )

        result = []
        for position in positions:
            result.append(
                PositionResponse(
                    id=position.id,  # type: ignore
                    symbol=position.symbol,  # type: ignore
                    quantity=position.quantity,  # type: ignore
                    average_price=position.average_price,  # type: ignore
                    current_price=position.current_price or 0.0,  # type: ignore
                    unrealized_pnl=position.unrealized_pnl,  # type: ignore
                )
            )

        return result
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{portfolio_id}/trades", response_model=List[TradeResponse])
async def get_portfolio_trades(
    portfolio_id: int, db: Session = Depends(get_db)
) -> List[TradeResponse]:
    """Get all trades for a portfolio"""
    try:
        trades = (
            db.query(Trade)
            .filter(
                Trade.portfolio_id == portfolio_id,
                Trade.user_id == 1,  # TODO: Use actual user_id
            )
            .all()
        )

        result = []
        for trade in trades:
            result.append(
                TradeResponse(
                    id=trade.id,  # type: ignore
                    symbol=trade.symbol,  # type: ignore
                    side=trade.side,  # type: ignore
                    order_type=trade.order_type,  # type: ignore
                    quantity=trade.quantity,  # type: ignore
                    price=trade.price or 0.0,  # type: ignore
                    filled_quantity=trade.filled_quantity,  # type: ignore
                    filled_price=trade.filled_price or 0.0,  # type: ignore
                    status=str(trade.status),
                    created_at=trade.created_at.isoformat() if trade.created_at else "",
                )
            )

        return result
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.put("/{portfolio_id}/update-positions")
async def update_portfolio_positions(
    portfolio_id: int, db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """Update portfolio positions with current market prices"""
    try:
        positions = (
            db.query(Position).filter(Position.portfolio_id == portfolio_id).all()
        )

        async with BitvavoAPI() as api:
            for position in positions:
                # Get current price for the symbol
                market = f"{position.symbol}-EUR"  # Assuming EUR as quote currency
                try:
                    ticker = await api.get_ticker(market)
                    current_price = float(ticker["last"])

                    # Update position with current price and calculate P&L
                    position.current_price = current_price  # type: ignore
                    position.unrealized_pnl = (  # type: ignore
                        current_price - position.average_price
                    ) * position.quantity

                except Exception as e:
                    print(f"Error updating price for {position.symbol}: {e}")
                    continue

        # Calculate total portfolio value
        portfolio = db.query(Portfolio).filter(Portfolio.id == portfolio_id).first()
        if portfolio:
            total_value = sum(
                (pos.current_price or 0) * pos.quantity for pos in positions
            )
            portfolio.total_value = total_value  # type: ignore

        db.commit()

        return {"message": "Portfolio positions updated successfully"}

    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/components/Portfolio.js
import React, { useState, useEffect } from "react";
import axios from "axios";

const Portfolio = () => {
  const [portfolios, setPortfolios] = useState([]);
  const [selectedPortfolio, setSelectedPortfolio] = useState(null);
  const [positions, setPositions] = useState([]);
  const [trades, setTrades] = useState([]);
  const [newPortfolioName, setNewPortfolioName] = useState("");
  const [loading, setLoading] = useState(true);
  const [message, setMessage] = useState("");

  useEffect(() => {
    fetchPortfolios();
  }, []);

  useEffect(() => {
    if (selectedPortfolio) {
      fetchPortfolioDetails(selectedPortfolio.id);
    }
  }, [selectedPortfolio]);

  const fetchPortfolios = async () => {
    try {
      setLoading(true);
      const response = await axios.get("/api/v1/portfolio/");
      setPortfolios(response.data);
      if (response.data.length > 0 && !selectedPortfolio) {
        setSelectedPortfolio(response.data[0]);
      }
    } catch (error) {
      console.error("Error fetching portfolios:", error);
      setMessage(
        "Error fetching portfolios: " +
          (error.response?.data?.detail || error.message)
      );
    } finally {
      setLoading(false);
    }
  };

  const fetchPortfolioDetails = async (portfolioId) => {
    try {
      const [positionsResponse, tradesResponse] = await Promise.all([
        axios.get(`/api/v1/portfolio/${portfolioId}/positions`),
        axios.get(`/api/v1/portfolio/${portfolioId}/trades`),
      ]);

      setPositions(positionsResponse.data);
      setTrades(tradesResponse.data);
    } catch (error) {
      console.error("Error fetching portfolio details:", error);
    }
  };

  const createPortfolio = async (e) => {
    e.preventDefault();
    if (!newPortfolioName.trim()) return;

    try {
      const response = await axios.post("/api/v1/portfolio/", {
        name: newPortfolioName,
        description: `Portfolio created on ${new Date().toLocaleDateString()}`,
      });

      setPortfolios([...portfolios, response.data]);
      setNewPortfolioName("");
      setMessage("Portfolio created successfully!");

      // Select the new portfolio
      setSelectedPortfolio(response.data);
    } catch (error) {
      setMessage(
        "Error creating portfolio: " +
          (error.response?.data?.detail || error.message)
      );
    }
  };

  const updatePortfolioPositions = async () => {
    if (!selectedPortfolio) return;

    try {
      await axios.put(
        `/api/v1/portfolio/${selectedPortfolio.id}/update-positions`
      );
      setMessage("Portfolio positions updated with current market prices!");
      fetchPortfolioDetails(selectedPortfolio.id);
      fetchPortfolios(); // Refresh to get updated total value
    } catch (error) {
      setMessage(
        "Error updating positions: " +
          (error.response?.data?.detail || error.message)
      );
    }
  };

  const formatPrice = (price) => {
    return new Intl.NumberFormat("nl-NL", {
      style: "currency",
      currency: "EUR",
      minimumFractionDigits: 2,
      maximumFractionDigits: 6,
    }).format(price);
  };

  const formatDate = (dateString) => {
    return new Date(dateString).toLocaleDateString("nl-NL", {
      year: "numeric",
      month: "short",
      day: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  };

  if (loading) return <div className="loading">Loading portfolios...</div>;

  return (
    <div>
      <h1>Portfolio Management</h1>

      {message && (
        <div
          className={message.includes("Error") ? "error" : "success"}
          style={{ marginBottom: "20px" }}
        >
          {message}
        </div>
      )}

      <div className="trading-dashboard">
        <div className="card">
          <h2>Portfolios</h2>

          <form onSubmit={createPortfolio} style={{ marginBottom: "20px" }}>
            <div className="form-group">
              <label>Create New Portfolio:</label>
              <input
                type="text"
                value={newPortfolioName}
                onChange={(e) => setNewPortfolioName(e.target.value)}
                placeholder="Enter portfolio name"
                style={{ marginRight: "10px" }}
              />
              <button type="submit" className="btn btn-primary">
                Create
              </button>
            </div>
          </form>

          {portfolios.length > 0 ? (
            <div>
              <div style={{ marginBottom: "15px" }}>
                <label>Select Portfolio: </label>
                <select
                  value={selectedPortfolio?.id || ""}
                  onChange={(e) => {
                    const portfolio = portfolios.find(
                      (p) => p.id === parseInt(e.target.value)
                    );
                    setSelectedPortfolio(portfolio);
                  }}
                  style={{ marginLeft: "10px", padding: "5px" }}
                >
                  {portfolios.map((portfolio) => (
                    <option key={portfolio.id} value={portfolio.id}>
                      {portfolio.name} - {formatPrice(portfolio.total_value)}
                    </option>
                  ))}
                </select>
              </div>

              <table className="market-table">
                <thead>
                  <tr>
                    <th>Portfolio</th>
                    <th>Total Value</th>
                    <th>Status</th>
                    <th>Description</th>
                  </tr>
                </thead>
                <tbody>
                  {portfolios.map((portfolio) => (
                    <tr
                      key={portfolio.id}
                      style={{
                        backgroundColor:
                          selectedPortfolio?.id === portfolio.id
                            ? "#404040"
                            : "transparent",
                        cursor: "pointer",
                      }}
                      onClick={() => setSelectedPortfolio(portfolio)}
                    >
                      <td>
                        <strong>{portfolio.name}</strong>
                      </td>
                      <td>{formatPrice(portfolio.total_value)}</td>
                      <td
                        className={
                          portfolio.is_active
                            ? "price-positive"
                            : "price-negative"
                        }
                      >
                        {portfolio.is_active ? "Active" : "Inactive"}
                      </td>
                      <td>{portfolio.description}</td>
                    </tr>
                  ))}
                </tbody>
              </table>
            </div>
          ) : (
            <p>No portfolios found. Create your first portfolio above.</p>
          )}
        </div>

        <div className="card">
          <h2>Portfolio Summary</h2>
          {selectedPortfolio ? (
            <div>
              <h3>{selectedPortfolio.name}</h3>
              <p>
                <strong>Total Value:</strong>{" "}
                {formatPrice(selectedPortfolio.total_value)}
              </p>
              <p>
                <strong>Positions:</strong> {positions.length}
              </p>
              <p>
                <strong>Total Trades:</strong> {trades.length}
              </p>

              <button
                onClick={updatePortfolioPositions}
                className="btn btn-secondary"
                style={{ marginTop: "10px" }}
              >
                Update with Current Prices
              </button>
            </div>
          ) : (
            <p>Select a portfolio to view summary</p>
          )}
        </div>
      </div>

      {selectedPortfolio && (
        <>
          <div className="card">
            <h2>Current Positions</h2>
            {positions.length > 0 ? (
              <div className="market-data">
                <table className="market-table">
                  <thead>
                    <tr>
                      <th>Symbol</th>
                      <th>Quantity</th>
                      <th>Avg Price</th>
                      <th>Current Price</th>
                      <th>Value</th>
                      <th>P&L</th>
                      <th>P&L %</th>
                    </tr>
                  </thead>
                  <tbody>
                    {positions.map((position) => {
                      const totalValue =
                        position.current_price * position.quantity;
                      const pnlPercent =
                        ((position.current_price - position.average_price) /
                          position.average_price) *
                        100;

                      return (
                        <tr key={position.id}>
                          <td>
                            <strong>{position.symbol}</strong>
                          </td>
                          <td>{position.quantity.toFixed(6)}</td>
                          <td>{formatPrice(position.average_price)}</td>
                          <td>{formatPrice(position.current_price)}</td>
                          <td>{formatPrice(totalValue)}</td>
                          <td
                            className={
                              position.unrealized_pnl >= 0
                                ? "price-positive"
                                : "price-negative"
                            }
                          >
                            {formatPrice(position.unrealized_pnl)}
                          </td>
                          <td
                            className={
                              pnlPercent >= 0
                                ? "price-positive"
                                : "price-negative"
                            }
                          >
                            {pnlPercent >= 0 ? "+" : ""}
                            {pnlPercent.toFixed(2)}%
                          </td>
                        </tr>
                      );
                    })}
                  </tbody>
                </table>
              </div>
            ) : (
              <p>
                No positions in this portfolio. Start trading to see positions
                here.
              </p>
            )}
          </div>

          <div className="card">
            <h2>Trade History</h2>
            {trades.length > 0 ? (
              <div className="market-data">
                <table className="market-table">
                  <thead>
                    <tr>
                      <th>Date</th>
                      <th>Symbol</th>
                      <th>Side</th>
                      <th>Type</th>
                      <th>Quantity</th>
                      <th>Price</th>
                      <th>Value</th>
                      <th>Status</th>
                    </tr>
                  </thead>
                  <tbody>
                    {trades.slice(0, 20).map((trade) => (
                      <tr key={trade.id}>
                        <td>{formatDate(trade.created_at)}</td>
                        <td>{trade.symbol}</td>
                        <td
                          className={
                            trade.side === "buy"
                              ? "price-positive"
                              : "price-negative"
                          }
                        >
                          {trade.side.toUpperCase()}
                        </td>
                        <td>{trade.order_type}</td>
                        <td>{trade.quantity.toFixed(6)}</td>
                        <td>
                          {trade.price ? formatPrice(trade.price) : "Market"}
                        </td>
                        <td>
                          {trade.price
                            ? formatPrice(trade.quantity * trade.price)
                            : "-"}
                        </td>
                        <td>{trade.status}</td>
                      </tr>
                    ))}
                  </tbody>
                </table>
              </div>
            ) : (
              <p>No trades in this portfolio yet.</p>
            )}
          </div>
        </>
      )}
    </div>
  );
};

export default Portfolio;
